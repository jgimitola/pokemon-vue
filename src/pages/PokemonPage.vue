<script>
import PokemonPicture from "../components/PokemonPicture.vue";
import PokemonOptions from "../components/PokemonOptions.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions.js";

export default {
  name: "PokemonPage",
  components: { PokemonPicture, PokemonOptions },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showing: false,
      message: "",
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const index = Math.floor(4 * Math.random());
      this.pokemon = this.pokemonArr[index];
    },
    handleSelection(pokemon) {
      this.showing = true;
      if (pokemon.id === this.pokemon.id) {
        this.message = "Has acertado!";
      } else {
        this.message = `Has fallado ;(. La respuesta  era ${this.pokemon.name}`;
      }
    },
    handleNewGame() {
      this.pokemonArr = [];
      this.pokemon = null;
      this.showing = false;
      this.message = "";

      this.mixPokemonArray();
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<template>
  <div v-if="!pokemon">Cargando...</div>

  <div v-else class="container">
    <h1>¿¡Quién es ese Pokemón!?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showing" />
    <PokemonOptions :pokemons="pokemonArr" @selection="handleSelection" />

    <template v-if="showing">
      <h2>{{ message }}</h2>
      <button @click="handleNewGame">Nuevo juego</button>
    </template>
  </div>
</template>

<style scoped>
.container {
  display: flex;
  flex-flow: nowrap column;
  align-items: center;
  gap: 1rem;
}
</style>
