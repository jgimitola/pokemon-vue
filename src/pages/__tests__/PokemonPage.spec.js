import { shallowMount } from "@vue/test-utils";
import { describe, expect, test, vitest } from "vitest";

import { pokemons } from "@/helpers/__tests__/mocks/pokemons.mock.js";
import PokemonPage from "@/pages/PokemonPage.vue";

describe("PokemonPage", () => {
  test("call mixPokemonArray on mount", () => {
    const mixPokemonArraySpy = vitest.spyOn(
      PokemonPage.methods,
      "mixPokemonArray"
    );
    shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("match with snapshot", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showing: false,
          message: "",
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("children custom components should exist", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showing: false,
          message: "",
        };
      },
    });

    const picture = wrapper.find("pokemon-picture-stub ");
    const options = wrapper.find("pokemon-options-stub");

    expect(picture.exists()).toBeTruthy();
    expect(picture.attributes("pokemonid")).toBe(`${pokemons[0].id}`);

    expect(options.exists()).toBeTruthy();
    expect(options.attributes("pokemons")).toBeDefined();
  });

  test("handleSelection works", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showing: false,
          message: "",
        };
      },
    });

    await wrapper.vm.handleSelection(pokemons[0]);

    expect(wrapper.vm.message).toBe("Has acertado!");

    const messageElement = wrapper.find("h2");

    expect(messageElement.text()).toBe("Has acertado!");
  });
});
