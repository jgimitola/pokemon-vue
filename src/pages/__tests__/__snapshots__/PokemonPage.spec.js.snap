// Vitest Snapshot v1

exports[`PokemonPage > match with snapshot 1`] = `
"<div data-v-ceac4f0d=\\"\\" class=\\"container\\">
  <h1 data-v-ceac4f0d=\\"\\">¿¡Quién es ese Pokemón!?</h1>
  <pokemon-picture-stub data-v-ceac4f0d=\\"\\" pokemonid=\\"1\\" showpokemon=\\"false\\"></pokemon-picture-stub>
  <pokemon-options-stub data-v-ceac4f0d=\\"\\" pokemons=\\"[object Object],[object Object],[object Object],[object Object]\\"></pokemon-options-stub>
  <!--v-if-->
</div>"
`;
